File: core/src/main/java/com/zp4rker/iab/listeners/PlayerJoinListener.java
package com.zp4rker.iab.listeners;

import com.zp4rker.iab.IABCore;
import com.zp4rker.iab.api.Explorer;
import com.zp4rker.iab.prompts.ExplorerSetupPrompt;
import org.bukkit.conversations.ConversationFactory;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;

public class PlayerJoinListener implements Listener {
    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        if (Explorer.find(event.getPlayer()) == null) {
            event.setJoinMessage(null);
            new ConversationFactory(IABCore.INST)
                    .withFirstPrompt(new ExplorerSetupPrompt())
                    .withModality(true)
                    .withLocalEcho(false)
                    .buildConversation(event.getPlayer()).begin();
        }

        Explorer explorer = Explorer.find(event.getPlayer());
        if (explorer != null) {
            event.getPlayer().setCustomName(explorer.getName());
            event.getPlayer().setDisplayName(explorer.getName());
            event.getPlayer().setPlayerListName(explorer.getName());
        }
    }
}

File: core/src/main/kotlin/com/zp4rker/iab/listeners/PlayerJoinListener.kt
package com.zp4rker.iab.listeners

import com.zp4rker.bukkot.extensions.runTask
import com.zp4rker.iab.LOGGER
import com.zp4rker.iab.PLUGIN
import com.zp4rker.iab.api.Explorer
import com.zp4rker.iab.prompts.ExplorerInitPrompt
import com.zp4rker.iab.utils.MM
import org.bukkit.conversations.ConversationFactory
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent

class PlayerJoinListener : Listener {
    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        if (Explorer.find(event.player) != null) return

        event.joinMessage(null)
        PLUGIN.runTask(async = true) { ExplorerInitPrompt(event.player).run() }
    }
}